### PR TITLE
Integrates RDS Analyzer rules engine into the MCP

### PR DESCRIPTION
**Pull request by mvazquezc**

### CODE REVIEW

File: pkg/mcpserver/rds_compare.go [hunk 1]
```diff
@@ -135,6 +153,12 @@ func HandleValidateRDS(ctx context.Context, req *mcp.CallToolRequest, input Vali
 		"validated", rdsResult.Validated,
 	)
 
+	// When analysis is enabled, force JSON output so the analyzer can parse the comparison
+	if input.RDSAnalysis {
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go]: **Concern: Silent override of user's `output_format`**

When `rds_analysis=true`, this silently replaces the user's requested `output_format` with `"json"`. The override is only logged at `Debug` level, so in production the user has no visibility into why their format choice was ignored. An LLM client requesting YAML will receive JSON with no indication why.

Would it make sense to return a `ValidationError` instead, similar to the `context`-requires-`kubeconfig` check at line 104? Something like:

```go
if input.RDSAnalysis && input.OutputFormat != "" && input.OutputFormat != "json" {
    err := NewValidationError("output_format",
        fmt.Sprintf("RDS analysis requires JSON output format, but '%s' was requested", input.OutputFormat),
        "Either omit output_format (defaults to JSON) or set it to 'json' when using rds_analysis")
    return newToolResultError(formatErrorForUser(err)), ValidateRDSOutput{}, nil
}
```

This way the user gets clear, actionable feedback rather than a surprise format change.

File: pkg/mcpserver/rds_compare.go [hunk 2]
```diff
@@ -3,30 +3,41 @@
...
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/rest"
 )
 
 // ValidateRDSResult is the structured response for the kube_compare_validate_rds tool.
 type ValidateRDSResult struct {
 	RDSReference *ResolveRDSResult `json:"rds_reference"`
 	Comparison   json.RawMessage   `json:"comparison"`
+	RDSAnalysis  string            `json:"rds_analysis,omitempty"`
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go]: **Concern: `rds_analysis` field mixes success output and error strings**

This field serves double duty — it contains analysis output (HTML/text/reporting) on success, or a raw Go error chain like `"Analysis failed: failed to fetch analysis rules: in-cluster config not available..."` on failure. A couple of concerns:

1. Consumers can't programmatically distinguish success from failure without string-matching on `"Analysis failed: "`.
2. The error message bypasses the project's `formatErrorForUser()` system (used everywhere else in the handler).
3. No content-type indicator — the field could contain HTML, plain text, or reporting text.

A possible approach would be to separate success and error into distinct fields:

```go
type ValidateRDSResult struct {
    RDSReference     *ResolveRDSResult `json:"rds_reference"`
    Comparison       json.RawMessage   `json:"comparison"`
    RDSAnalysis      string            `json:"rds_analysis,omitempty"`
    RDSAnalysisError string            `json:"rds_analysis_error,omitempty"`
}
```

This way consumers can check for errors structurally. What do you think?

**Comment by sakhoury** [comment]
**Concern: Tool description and README don't mention the new analysis capability**

The `Description` string at `rds_compare.go:50` is the primary signal LLM clients use to discover and decide when to invoke a tool ([MCP spec 2025-11-25](https://modelcontextprotocol.io/specification/2025-11-25/server/tools)). Without mentioning analysis here, no LLM client will know to set `rds_analysis: true`.

A possible update:

```go
Description: "Validate an OpenShift cluster's compliance with Red Hat Telco RDS. " +
    "Optionally classifies deviations using RDS Analyzer rules into impact categories " +
    "(Impacting, NotImpacting, NeedsReview, NotADeviation) when rds_analysis is enabled. " +
    "This is the recommended tool for RDS validation.",
```

Also, the README's parameter table and response example for this tool (`README.md` lines 347-371) should be updated to document the new `rds_analysis` and `rds_analysis_format` input parameters and the `rds_analysis` output field. An example prompt like *"Validate my cluster against the Telco Core RDS and analyze the impact of any deviations"* would help users discover the feature too.

File: pkg/mcpserver/rds_compare.go [hunk 3]
```diff
@@ -168,6 +192,20 @@ func HandleValidateRDS(ctx context.Context, req *mcp.CallToolRequest, input Vali
 		Comparison:   comparisonJSON,
 	}
 
+	if input.RDSAnalysis {
+		analysisFormat := input.RDSAnalysisFormat
+		if analysisFormat == "" {
+			analysisFormat = "html"
+		}
+		analysisOutput, analysisErr := RunRDSAnalysis(ctx, comparisonOutput, input.RDSType, rdsResult.ClusterVersion, analysisFormat, fetchAnalysisRulesFromConfigMap)
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go]: **Suggestion: Consider injecting `RulesFetcher` via a service struct**

The production fetcher `fetchAnalysisRulesFromConfigMap` is hardcoded here rather than injected. The rest of the codebase uses dependency injection via service structs (`CompareService`, `ReferenceService`) and interfaces defined in `interfaces.go` with `//go:generate mockgen` directives.

This makes the `HandleValidateRDS` integration path untestable without a real in-cluster Kubernetes config. Additionally, `fetchAnalysisRulesFromConfigMap` creates a new Kubernetes client on every invocation (`rest.InClusterConfig()` + `dynamic.NewForConfig()`) with no caching.

It might be worth storing `RulesFetcher` on a service struct or as a configurable package-level variable (following the `DefaultRegistry`/`DefaultClusterFactory` pattern in `interfaces.go`) so the handler integration can be tested with a mock fetcher.

File: pkg/mcpserver/rds_compare.go [hunk 4]
```diff
@@ -49,13 +60,20 @@ func ValidateRDSTool() *mcp.Tool {
...
-	Context      string
-	RDSType      string
-	OutputFormat string
-	AllResources bool
+	Kubeconfig        string
+	Context           string
+	RDSType           string
+	OutputFormat      string
+	AllResources      bool
+	RDSAnalysis       bool
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go]: **Nit: These new fields appear unused in production code**

`HandleValidateRDS` uses `ValidateRDSInput` directly and never constructs or reads from `ValidateRDSArgs`. These two new fields seem to be dead code — they're only exercised in a test that verifies struct creation.

Would it make sense to remove them until they're actually consumed, to avoid maintaining fields that could drift out of sync?

**Comment by sakhoury** [comment]
**Suggestion: Integration test for `HandleValidateRDS` with `rds_analysis=true`**

The tests cover `RunRDSAnalysis` in isolation nicely — great job on the test fixtures and the `RulesFetcher` injection pattern! The handler integration path though (format override, non-fatal error handling, and default format fallback at `rds_compare.go:195-207`) isn't tested yet.

Once the `RulesFetcher` is injectable (per the DI comment on the inline review), it would be great to add tests for:
- Analysis enabled with a mock fetcher — verify response includes analysis output
- Analysis failure producing a non-fatal result — verify comparison still succeeds with error in response
- Analysis disabled — verify no `rds_analysis` field in output

**Comment by sakhoury** [comment]
**Minor: New input parameters missing from debug logging**

The "Parsed arguments" debug log at `rds_compare.go:128-134` doesn't include the new `rdsAnalysis` and `rdsAnalysisFormat` fields. Adding them would help with debugging, especially when tracing why the output format was overridden.

```go
logger.Debug("Parsed kube_compare_validate_rds arguments",
    "rdsType", input.RDSType,
    "hasKubeconfig", kubeconfig != "",
    "context", input.Context,
    "outputFormat", input.OutputFormat,
    "allResources", input.AllResources,
    "rdsAnalysis", input.RDSAnalysis,           // add
    "rdsAnalysisFormat", input.RDSAnalysisFormat, // add
)
```

File: pkg/mcpserver/rds_compare.go [hunk 5]
```diff
@@ -168,6 +192,20 @@ func HandleValidateRDS(ctx context.Context, req *mcp.CallToolRequest, input Vali
 		Comparison:   comparisonJSON,
 	}
 
+	if input.RDSAnalysis {
+		analysisFormat := input.RDSAnalysisFormat
+		if analysisFormat == "" {
+			analysisFormat = "html"
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go]: **Nit: Default `"html"` is duplicated**

The default analysis format `"html"` is set in the schema (`schema.go:72`) and again as a runtime fallback here. Consider extracting to a named constant to keep them in sync and prevent drift.

File: pkg/mcpserver/rds_compare.go [hunk 6]
```diff
@@ -184,3 +222,82 @@ func HandleValidateRDS(ctx context.Context, req *mcp.CallToolRequest, input Vali
...
+	if err != nil {
+		return nil, fmt.Errorf("in-cluster config not available for rules ConfigMap: %w", err)
+	}
+
+	client, err := dynamic.NewForConfig(inClusterConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create dynamic client for rules ConfigMap: %w", err)
+	}
+
+	cm, err := client.Resource(configMapGVR).Namespace(DefaultReferenceConfigNamespace).Get(ctx, rdsAnalyzerRulesConfigMap, metav1.GetOptions{})
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go:269]: **Note: `DefaultReferenceConfigNamespace` is scoped to BIOS in its comment**

`configMapGVR` and `DefaultReferenceConfigNamespace` are defined in `bios_compare.go` with the comment *"default namespace for BIOS reference ConfigMaps"*. Now they're also used for RDS analyzer rules, which is a different concept.

Might be worth either moving the shared constants to a common file (e.g., `constants.go`), or updating the comment in `bios_compare.go` to reflect the broader usage. Just a thought to keep things discoverable as the codebase grows!

File: pkg/mcpserver/rds_compare.go [hunk 7]
```diff
@@ -184,3 +222,82 @@ func HandleValidateRDS(ctx context.Context, req *mcp.CallToolRequest, input Vali
...
+	}
+	if !found || data == "" {
+		return nil, fmt.Errorf("key %q not found in ConfigMap %s/%s", dataKey, DefaultReferenceConfigNamespace, rdsAnalyzerRulesConfigMap)
+	}
+
+	return []byte(data), nil
+}
+
+// RunRDSAnalysis runs the RDS impact analysis on comparison JSON output.
+func RunRDSAnalysis(ctx context.Context, comparisonJSON, rdsType, clusterVersion, analysisFormat string, fetchRules RulesFetcher) (string, error) {
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go:287]: **Suggestion: Defensive validation for empty `clusterVersion`**

Since `RunRDSAnalysis` is exported, it could be called with an empty `clusterVersion`. In that case, `ExtractMajorMinorVersion("")` returns `"v"` and `TrimPrefix` produces `""`, passing an empty string to the analyzer.

A small guard would make this more robust:

```go
if clusterVersion == "" {
    return "", fmt.Errorf("cluster version is required for RDS analysis")
}
```

File: pkg/mcpserver/rds_compare_test.go [hunk 8]
```diff
@@ -26,17 +75,96 @@ var _ = Describe("RDSCompareHandler", func() {
...
+			prop := schema.Properties["rds_analysis_format"]
+			Expect(prop.Default).To(Equal(json.RawMessage(`"html"`)))
+		})
+	})
+
+	Describe("rulesKeyForRDSType", func() {
+		It("returns core-rules.yaml for core type", func() {
+			Expect(mcpserver.RulesKeyForRDSType("core")).To(Equal("core-rules.yaml"))
+		})
+
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare_test.go:100]: **Suggestion: Consider adding a test for the `"hub"` RDS type**

Tests cover `"core"` and `"ran"` but not `"hub"`, which was recently added. It's cheap to verify:

```go
It("returns hub-rules.yaml for hub type", func() {
    Expect(mcpserver.RulesKeyForRDSType("hub")).To(Equal("hub-rules.yaml"))
})
```

File: pkg/mcpserver/rds_compare_test.go [hunk 9]
```diff
@@ -26,17 +75,96 @@ var _ = Describe("RDSCompareHandler", func() {
...
+			})
+		})
+
+		Context("with rules fetch failure", func() {
+			It("returns an error", func() {
+				failingFetcher := func(_ context.Context, _ string) ([]byte, error) {
+					return nil, fmt.Errorf("configmap not found")
+				}
+				_, err := mcpserver.RunRDSAnalysis(context.Background(), testComparisonJSON, "core", "4.20.0", "html", failingFetcher)
+				Expect(err).To(HaveOccurred())
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare_test.go:167]: **Suggestion: Test with malformed rules YAML**

There's no test where the fetcher returns syntactically invalid YAML. This would exercise the `"failed to initialize analyzer"` error path. A quick addition:

```go
Context("with invalid rules YAML", func() {
    It("returns an error", func() {
        badFetcher := func(_ context.Context, _ string) ([]byte, error) {
            return []byte("not: valid: rules: yaml: !!!"), nil
        }
        _, err := mcpserver.RunRDSAnalysis(context.Background(), testComparisonJSON, "core", "4.20.0", "html", badFetcher)
        Expect(err).To(HaveOccurred())
        Expect(err.Error()).To(ContainSubstring("failed to initialize analyzer"))
    })
})
```

File: pkg/mcpserver/rds_compare_test.go [hunk 10]
```diff
@@ -26,17 +75,96 @@ var _ = Describe("RDSCompareHandler", func() {
...
+			It("produces HTML output", func() {
+				output, err := mcpserver.RunRDSAnalysis(context.Background(), testComparisonJSON, "core", "4.20.0", "html", validFetcher)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(output).To(ContainSubstring("<!DOCTYPE html>"))
+			})
+
+			It("produces reporting output", func() {
+				output, err := mcpserver.RunRDSAnalysis(context.Background(), testComparisonJSON, "core", "4.20.0", "reporting", validFetcher)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(output).NotTo(BeEmpty())
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare_test.go]: **Suggestion: Test for default/fallback format behavior**

The `default` case in `RunRDSAnalysis` silently falls back to HTML for any unrecognized format (including empty string). Since `RunRDSAnalysis` is exported, it would be good to verify this behavior:

```go
It("falls back to HTML for empty format", func() {
    output, err := mcpserver.RunRDSAnalysis(ctx, testComparisonJSON, "core", "4.20.0", "", validFetcher)
    Expect(err).NotTo(HaveOccurred())
    Expect(output).To(ContainSubstring("<!DOCTYPE html>"))
})
```

File: pkg/mcpserver/rds_compare_test.go [hunk 11]
```diff
@@ -26,17 +75,96 @@ var _ = Describe("RDSCompareHandler", func() {
...
+	Describe("RunRDSAnalysis", func() {
+		validFetcher := func(_ context.Context, _ string) ([]byte, error) {
+			return []byte(testAnalysisRulesYAML), nil
+		}
+
+		Context("with valid rules and comparison JSON", func() {
+			It("produces non-empty text output", func() {
+				output, err := mcpserver.RunRDSAnalysis(context.Background(), testComparisonJSON, "core", "4.20.0", "text", validFetcher)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(output).NotTo(BeEmpty())
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare_test.go]: **Suggestion: Strengthen output assertions**

The text and reporting format tests only check `NotTo(BeEmpty())`, which would pass even if the analyzer returned unexpected output. Consider asserting on a substring that confirms the diff was actually processed, e.g. the CR name `"v1_ConfigMap_default_test"` appearing in the output, or a classification like `"NotImpacting"`.

File: pkg/mcpserver/rds_compare.go [hunk 12]
```diff
@@ -184,3 +222,82 @@ func HandleValidateRDS(ctx context.Context, req *mcp.CallToolRequest, input Vali
...
+	rulesData, err := fetchRules(ctx, rdsType)
+	if err != nil {
+		return "", fmt.Errorf("failed to fetch analysis rules: %w", err)
+	}
+
+	// Bridge version format: ExtractMajorMinorVersion returns "v4.20",
+	// but rds-analyzer's ParseOCPVersion expects "4.20"
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go]: **Suggestion: Simplify version bridging comment**

This comment references `ParseOCPVersion`, which is an internal function of rds-analyzer. If the library refactors its internals, the comment becomes misleading while the code might still be correct.

A simpler version that focuses on the "why" without coupling to library internals:

```go
// rds-analyzer expects version without "v" prefix (e.g., "4.20" not "v4.20")
```

File: pkg/mcpserver/rds_compare.go (hunk 1 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go]: Makes sense. Changed.

File: pkg/mcpserver/rds_compare.go (hunk 2 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go]: Implemented this.

File: pkg/mcpserver/rds_compare.go (hunk 3 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go]: Done.

File: pkg/mcpserver/rds_compare.go (hunk 4 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go]: Yes. I just removed it.

File: pkg/mcpserver/rds_compare.go (hunk 5 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go]: Moved to a constant as suggested.

File: pkg/mcpserver/rds_compare.go (hunk 6 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go:269]: Moved them to constants.go file.

File: pkg/mcpserver/rds_compare.go (hunk 7 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go:287]: added defensive validation and the respective test.

File: pkg/mcpserver/rds_compare_test.go (hunk 8 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare_test.go:100]: Added.

File: pkg/mcpserver/rds_compare_test.go (hunk 9 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare_test.go:167]: Added test, I used the same test content used in RDSAnalyzer library.

`return []byte("not: [valid: yaml"), nil`

File: pkg/mcpserver/rds_compare_test.go (hunk 10 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare_test.go]: Added.

File: pkg/mcpserver/rds_compare_test.go (hunk 11 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare_test.go]: Added checks with the content of the test rule.

File: pkg/mcpserver/rds_compare.go (hunk 12 above)
Comment (mvazquezc) [reply pkg/mcpserver/rds_compare.go]: Changed.

File: pkg/mcpserver/rds_compare.go [hunk 13]
```diff
@@ -47,15 +62,37 @@ func ValidateRDSTool() *mcp.Tool {
...
+
+var defaultAnalysisService = NewAnalysisService()
+
+// Package-level function variables for testability.
+// Tests can swap these to inject fakes without changing the handler signature.
+var (
+	resolveRDSFunc        = ResolveRDSInternal
+	validateReferenceFunc = validateReference
+	runCompareFunc        = RunCompare
+)
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go:94]: The three new function variables (`resolveRDSFunc`, `validateReferenceFunc`, `runCompareFunc`) introduce a different testability mechanism from what the codebase uses everywhere else. The existing pattern is interface-based DI via service structs (`CompareService`, `ReferenceService`) with mockgen in `interfaces.go`, tested from `package mcpserver_test` (black-box). This PR instead uses mutable package-level variables swapped in a white-box internal test file (`package mcpserver`).

A few things stand out:

1. The internal test file contradicts the project's convention of black-box testing via `package mcpserver_test` (every other test file in the package uses this).
2. The save/restore pattern in `BeforeEach`/`AfterEach` is fragile - if a test panics between the two, the globals stay corrupted for the rest of the suite.
3. These variables serve no purpose in production. They exist solely for the internal test file, which creates a circular dependency between the two.

The external test file already provides thorough coverage of `RunRDSAnalysis` with all format variations, error paths, and edge cases. The handler integration tests in the internal file mostly duplicate that coverage through a different mechanism.

Would it make sense to either:

(a) Make `HandleValidateRDS` a method on a service struct (consistent with how `ReferenceService.ResolveRDS` works), allowing the handler to be tested via gomock from `package mcpserver_test`, or

(b) Remove the function variables and the internal test file entirely, relying on the existing external tests for `RunRDSAnalysis` (which are already comprehensive)?

Option (a) is the cleaner long-term approach. Option (b) is less churn for this PR.

**Review by sakhoury — COMMENTED** [review]
All 12 comments from the first round have been addressed - nice work. A few more things came up while going through the updated code.

File: pkg/mcpserver/errors.go [hunk 14]
```diff
@@ -192,6 +195,12 @@ func FormatErrorForUser(err error) string {
 			"Please use token, client certificate, or OIDC authentication instead."
 	}
 
+	if errors.Is(err, ErrRDSAnalysisFailed) {
+		return "RDS deviation analysis could not be completed. " +
+			"Ensure the MCP server is running in-cluster and the analysis rules ConfigMap is deployed. " +
+			"The comparison results are still valid."
+	}
```
Comment (sakhoury) [inline pkg/mcpserver/errors.go:202]: When `ErrRDSAnalysisFailed` is matched, `FormatErrorForUser` returns a fixed string and throws away the underlying error entirely. This means five very different failure modes all produce the same generic message:

- "in-cluster config not available" (server not in a pod)
- "configmap not found" (rules not deployed)
- "key 'hub-rules.yaml' not found" (wrong rds_type key in the ConfigMap)
- "failed to initialize analyzer" (malformed rules YAML)
- "analysis failed" (rds-analyzer internal error)

The other sentinel errors in this function handle security-sensitive kubeconfig data, so hiding details there makes sense. But analysis failures are purely operational and safe to expose. The consumer here is an AI assistant that can reason about technical error messages, so giving it the actual cause would be more helpful than a generic "ensure the MCP server is running in-cluster" message.

The original error is logged at Warn level, but the user-facing `rds_analysis_error` field has no diagnostic value at all right now.

Something like this would preserve the friendly framing while including the cause:

```go
if errors.Is(err, ErrRDSAnalysisFailed) {
    cause := errors.Unwrap(err)
    return fmt.Sprintf("RDS deviation analysis could not be completed: %v. "+
        "The comparison results are still valid.", cause)
}
```

This way an operator (or an LLM) can tell the difference between a deployment issue and a configuration problem without needing access to server logs.

File: pkg/mcpserver/rds_compare.go [hunk 15]
```diff
@@ -135,6 +174,15 @@ func HandleValidateRDS(ctx context.Context, req *mcp.CallToolRequest, input Vali
...
 	)
 
+	// When analysis is enabled, JSON output is required, if the client requested a different output
+	// throw a validation error.
+	if input.RDSAnalysis && input.OutputFormat != "" && input.OutputFormat != "json" {
+		err := NewValidationError("output_format",
+			fmt.Sprintf("RDS analysis requires JSON output format, but '%s' was requested", input.OutputFormat),
+			"Either omit output_format (defaults to JSON) or set it to 'json' when using rds_analysis")
+		return newToolResultError(formatErrorForUser(err)), ValidateRDSOutput{}, nil
+	}
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go:184]: The check for incompatible `output_format` + `rds_analysis` currently runs after `resolveRDSFunc` has already executed. That means a user who passes `rds_analysis: true` with `output_format: "yaml"` pays the full cost of cluster version detection and registry lookups before getting a validation error that could have been caught immediately from the input alone.

The existing `context requires kubeconfig` validation at line 122 already follows the right pattern - input validation first, expensive calls second. Moving this check up to the same area would be consistent:

```go
// Near line 130, alongside other input validation
if input.RDSAnalysis && input.OutputFormat != "" && input.OutputFormat != "json" {
    err := NewValidationError("output_format",
        fmt.Sprintf("RDS analysis requires JSON output format, but '%s' was requested", input.OutputFormat),
        "Either omit output_format (defaults to JSON) or set it to 'json' when using rds_analysis")
    return newToolResultError(formatErrorForUser(err)), ValidateRDSOutput{}, nil
}
```

File: pkg/mcpserver/rds_compare_internal_test.go [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+			result, _, err := HandleValidateRDS(context.Background(), nil, input)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(result.IsError).To(BeTrue())
+
+			textContent, ok := result.Content[0].(*mcp.TextContent)
+			Expect(ok).To(BeTrue())
+			Expect(textContent.Text).To(ContainSubstring("JSON"))
+		})
+	})
+})
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare_internal_test.go:248]: All handler integration tests mock `resolveRDSFunc` to return `ClusterVersion: "4.20.0"`. The `RunRDSAnalysis` unit test covers the empty version guard at the function level, but there is no handler-level test verifying what happens when `ResolveRDSInternal` returns an empty `ClusterVersion` and `rds_analysis` is true.

If `ResolveRDSInternal` ever returns a result with an empty version (e.g. from a cluster that does not report it properly), the handler should populate `RDSAnalysisError` gracefully rather than producing meaningless output. A quick test would cover this:

```go
Context("with RDS analysis enabled but empty cluster version from resolve", func() {
    BeforeEach(func() {
        resolveRDSFunc = func(_ context.Context, _ *ResolveRDSArgs) (*ResolveRDSResult, error) {
            return &ResolveRDSResult{
                ClusterVersion: "",
                Reference:      "https://example.com/metadata.yaml",
                Validated:      true,
            }, nil
        }
        defaultAnalysisService.FetchRules = func(_ context.Context, _ string) ([]byte, error) {
            return []byte(testRulesYAML), nil
        }
    })

    It("returns comparison results with analysis error, not a tool error", func() {
        input := ValidateRDSInput{RDSType: "core", RDSAnalysis: true}
        result, _, err := HandleValidateRDS(context.Background(), nil, input)
        Expect(err).NotTo(HaveOccurred())
        Expect(result.IsError).To(BeFalse())

        var parsed ValidateRDSResult
        // ... verify RDSAnalysisError is populated
    })
})
```

File: pkg/mcpserver/rds_compare.go [hunk 17]
```diff
@@ -47,15 +62,37 @@ func ValidateRDSTool() *mcp.Tool {
...
+const rdsAnalyzerRulesConfigMap = "rds-analyzer-rules"
+
+// analysisFormatText is the rds-analyzer format value for plain text output.
+const analysisFormatText = "text"
+
+// AnalysisService encapsulates dependencies for RDS analysis operations.
+// This enables dependency injection for testing.
+type AnalysisService struct {
+	FetchRules RulesFetcher
 }
```
Comment (sakhoury) [inline pkg/mcpserver/rds_compare.go:77]: `AnalysisService` currently has zero methods - it only holds a `FetchRules` field. Compare with `ReferenceService` (has `ResolveRDS`) and `CompareService` (has `ValidateHTTPReference`, `ValidateOCIReference`). The handler extracts `defaultAnalysisService.FetchRules` and passes it directly to `RunRDSAnalysis`, so the service struct is just a pass-through holder.

Additionally, `RulesFetcher` is defined as a function type rather than an interface, which diverges from the project convention where all DI points are interfaces in `interfaces.go` with `//go:generate mockgen` directives.

Two options depending on how much refactoring you want to do in this PR:

(a) Give `AnalysisService` a real method that wraps the analysis logic, and make `HandleValidateRDS` call it through the service. Define `RulesFetcher` as an interface in `interfaces.go` with a mockgen directive. This aligns fully with the existing patterns.

(b) Drop `AnalysisService` entirely and pass `RulesFetcher` as a parameter to `RunRDSAnalysis` (which is already the case). The service struct adds no value in its current form. This is the simpler option if you want to keep the scope of this PR focused.

Either way the end result is cleaner than the current middle ground where the service exists but does nothing.

**Comment by mvazquezc** [comment]
@midu16 let's sync on this when you're back.
